### PR TITLE
Addresses #15336 by preventing a notation to silently set variables at level strictly less than 0

### LINKED ISSUE #15336: level-0 infix notations in a custom entry are not used for printing
#### Description of the problem

```coq

Declare Custom Entry ent.
Notation "ent:( x )" := x (x custom ent at level 9999, at level 0, only parsing).
Notation "!" := (Set) (in custom ent at level 0).
Notation "a ; b" := (prod a b) (in custom ent at level 0, a custom ent, b custom ent).
Check ent:( ! ; ! ).
(* (ent:( !) * ent:( !))%type *)
Notation "a ;' b" := (prod a b) (in custom ent at level 1, a custom ent, b custom ent).
Check ent:( ! ;' ! ).
(* ent:( !;' !) *)
```
#### Coq Version

8.14.0

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** enhancement

The PR has the effect of letting the typical notations below to fail:
```coq
Notation "a ; b" := (prod a b) (at level 0).
(* To respect expected associativity, a would need to be set at level strictly less than 0. *)
Notation "[ a ; b ]" := (prod a b) (a at next level).
(* The notation being at level 0, a would need to be set at level strictly less than 0 *)
```

It also adds an explicit `right associativity` in notations at level 0 in the standard library or test suite (or use level 1 when left associativity is needed).

There will be probably many CI failures due to notations at level 0 not necessarily setting the associativity. We'll see if a special treatment is needed.

Closes #15336.

- [x] Added / updated **test-suite**.
- [x] Added **changelog**.

#### Overlay

- [ ] https://github.com/DeepSpec/sf/pull/10

### CODE REVIEW

File: test-suite/output/Notations4.v
Comment (gares) [inline test-suite/output/Notations4.v:267]: why don't you put the new test later? it would avoid the line changes in the out file.

Comment (herbelin) [reply test-suite/output/Notations4.v:267]: This is what I used to do but that created conflicts when two PR concurrenty added tests at the end. Then, I had the idea to add tests at some relevant places in the core of the file, avoiding the end, hoping for less conflicts. Unfortunately, this creates the line number conflict :( I'll go back to adding at the end...

Comment (SkySkimmer) [reply test-suite/output/Notations4.v:267]: Why not in a new file?

Comment (herbelin) [reply test-suite/output/Notations4.v:267]: > Why not in a new file?

That's also a possibility (maybe am I afraid it adds the overhead of interacting with the file system and calling coq again). Here, it would however have the advantage of being able to give both a bug number and a desription in the file name, say, `NotationLevelLessZero_bug15336.v` (though such a long name is not so common).

**Comment by ppedrot** [comment]
Doesn't seem to be ready, removing the milestone.

**Comment by herbelin** [comment]
@coqbot run full ci

**Comment by ppedrot** [comment]
@herbelin what's the status of this PR? IIUC this is essentially lacking overlays from the broken devs, isn't it?

**Comment by herbelin** [comment]
You're right, maybe four overlays is not too unreasonable. I thought I could turn the error into a warning otherwise.

**Comment by ppedrot** [comment]
@herbelin I think this PR should go in, but then it needs to be rebased and overlays must be written.

**Comment by herbelin** [comment]
Status:
- Rebased done.
- Current state is:
  - an error on left-associative notations at level 0 (impact to be tested in the CI).
  - a warning with automatic change of the level on right-associative notations at level 0 and on notations at level 0 referring to next level (this only needs an overlay for flocq, removing the "(at level 0)" in line 125 of src/IEEE754/Int63Copy.v).
- Waiting for Inria gitlab to restart before running the CI.

**Review by proux01 — APPROVED** [review]
Rebased, overlay added, will merge when CI is green.

**Comment by herbelin** [comment]
@proux01: I would prefer to postpone a bit this PR as I'm studying a less invasive change. E.g., the PR still introduces a failure of sf which is not justified.

**Comment by proux01** [comment]
Ok, otherwise what about merging the current PR which is already an improvement? That would let you plenty of time to improve without bothering users too much since this is now targetting 8.19 anyway.

**Comment by herbelin** [comment]
@proux01: I wrote  #17859 to explain my concerns with the approach taken in this PR. It is quite delicate...

**Pull request closed without merging**
